src/main/java/io/synadia/flink/v0/source/reader/NatsSubjectSplitReader.java

```diff
@@ -142,29 +151,71 @@ public void notifyCheckpointComplete(String subject, List<Message> messages)
         // For instance if we know it's ack all, we could look for the message
         // with the highest consumer sequence and just ack that one
 
-        // MANUAL ACK IS DONE INTENTIONALLY
-        // The message was received on a different connection,
-        // which could be closed at the time of ack
-        // TODO see if connection resources can be managed better
-        //noinspection resource
-        Connection conn = connection();
-        for (Message m : messages) {
-            conn.publish(m.getReplyTo(), ACK_BODY_BYTES);
+        contextLock.lock();
```

Comment:
Do we need to make this implementation thread-safe ? Is there a possibility of having a race condition, given this method is invoked from one place periodically ?